resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/Client.scala

```diff
@@ -224,121 +158,207 @@ private[spark] class Client(
+    val topLevelMessage = s"The driver pod with name ${driverPod.getMetadata.getName}" +
+      s" in namespace ${driverPod.getMetadata.getNamespace} was not ready in" +
+      s" $driverLaunchTimeoutSecs seconds."
+    val podStatusPhase = if (driverPod.getStatus.getPhase != null) {
+      s"Latest phase from the pod is: ${driverPod.getStatus.getPhase}"
+    } else {
+      "The pod had no final phase."
+    }
+    val podStatusMessage = if (driverPod.getStatus.getMessage != null) {
+      s"Latest message from the pod is: ${driverPod.getStatus.getMessage}"
+    } else {
+      "The pod had no final message."
+    }
+    val failedDriverContainerStatusString = driverPod.getStatus
+      .getContainerStatuses
+      .asScala
+      .find(_.getName == DRIVER_LAUNCHER_CONTAINER_NAME)
+      .map(status => {
+        val lastState = status.getState
+        if (lastState.getRunning != null) {
+          "Driver container last state: Running\n" +
+            s"Driver container started at: ${lastState.getRunning.getStartedAt}"
+        } else if (lastState.getWaiting != null) {
+          "Driver container last state: Waiting\n" +
+            s"Driver container wait reason: ${lastState.getWaiting.getReason}\n" +
+            s"Driver container message: ${lastState.getWaiting.getMessage}\n"
+        } else if (lastState.getTerminated != null) {
+          "Driver container last state: Terminated\n" +
+            s"Driver container started at: ${lastState.getTerminated.getStartedAt}\n" +
+            s"Driver container finished at: ${lastState.getTerminated.getFinishedAt}\n" +
+            s"Driver container exit reason: ${lastState.getTerminated.getReason}\n" +
+            s"Driver container exit code: ${lastState.getTerminated.getExitCode}\n" +
+            s"Driver container message: ${lastState.getTerminated.getMessage}"
+        } else {
+          "Driver container last state: Unknown"
+        }
+      }).getOrElse("The driver container wasn't found in the pod; expected to find" +
+      s" container with name $DRIVER_LAUNCHER_CONTAINER_NAME")
+    s"$topLevelMessage\n" +
+      s"$podStatusPhase\n" +
+      s"$podStatusMessage\n\n$failedDriverContainerStatusString"
+  }
 
-    addPortToServiceAndContainer(
-      UI_PORT_NAME,
-      sparkConf
-        .getOption("spark.ui.port")
-        .map(_.toInt)
-        .getOrElse(DEFAULT_UI_PORT))
-    (servicePorts, containerPorts)
+  private def buildContainerPorts(): Seq[ContainerPort] = {
+    Seq(
+      new ContainerPortBuilder()
+        .withContainerPort(sparkConf.getInt("spark.driver.port", DEFAULT_DRIVER_PORT))
+        .build(),
+      new ContainerPortBuilder()
+        .withContainerPort(sparkConf.getInt("spark.blockmanager.port", DEFAULT_BLOCKMANAGER_PORT))
+        .build(),
+      new ContainerPortBuilder()
+        .withContainerPort(DRIVER_LAUNCHER_SERVICE_INTERNAL_PORT)
+        .build(),
+      new ContainerPortBuilder()
+        .withContainerPort(uiPort)
+        .build())
```

Comment:
I think this can be rewritten to:

```
  private def buildContainerPorts(): Seq[ContainerPort] = {
    Seq(sparkConf.getInt("spark.driver.port", DEFAULT_DRIVER_PORT),
        sparkConf.getInt("spark.blockmanager.port", DEFAULT_BLOCKMANAGER_PORT),
        DRIVER_LAUNCHER_SERVICE_INTERNAL_PORT
        uiPort)
      .map(p => new ContainerPortBuilder().withContainerPort(p).build())
  }
```